src/execution.jl

```diff
@@ -64,24 +64,71 @@ end
 world_age() = ccall(:jl_get_tls_world_age, UInt, ())
 
 # slow lookup of local method age
-function method_age(f, tt)::UInt
-    for m in Base._methods(f, tt, 1, typemax(UInt))
-        @static if VERSION >= v"1.2.0-DEV.573"
+function method_age(f, t)::UInt
+    for m in Base._methods(f, t, 1, typemax(UInt))
+        if VERSION >= v"1.2.0-DEV.573"
             return m[3].primary_world
         else
             return m[3].min_world
         end
     end
+
+    tt = Base.to_tuple_type(t)
     throw(MethodError(f, tt))
 end
+## high-level @cuda interface
+
+"""
+    @cuda [kwargs...] func(args...)
+
+High-level interface for executing code on a GPU. The `@cuda` macro should prefix a call,
+with `func` a callable function or object that should return nothing. It will be compiled to
+a CUDA function upon first use, and to a certain extent arguments will be converted and
+managed automatically using `cudaconvert`. Finally, a call to `CUDAdrv.cudacall` is
+performed, scheduling a kernel launch on the current CUDA context.
+
+Several keyword arguments are supported that influence the behavior of `@cuda`.
+- `dynamic`: use dynamic parallelism to launch device-side kernels
+- arguments that influence kernel compilation: see [`cufunction`](@ref) and
+  [`dynamic_cufunction`](@ref)
+- arguments that influence kernel launch: see [`CUDAnative.HostKernel`](@ref) and
+  [`CUDAnative.DeviceKernel`](@ref)
+
+The underlying operations (argument conversion, kernel compilation, kernel call) can be
+performed explicitly when more control is needed, e.g. to reflect on the resource usage of a
+kernel to determine the launch configuration. A host-side kernel launch is done as follows:
+
+    args = ...
+    GC.@preserve args begin
+        kernel_args = cudaconvert.(args)
+        kernel_tt = Tuple{Core.Typeof.(kernel_args)...}
+        kernel = cufunction(f, kernel_tt; compilation_kwargs)
+        kernel(kernel_args...; launch_kwargs)
+    end
 
+A device-side launch, aka. dynamic parallelism, is similar but more restricted:
```

Comment:
We don't support device-side launch right now, but I suppose we can leave this for the future (AMD does support this functionality).